### PR TITLE
feat: reject suspicious x:expect[attribute(test)]

### PR DESCRIPTION
**Pull request by AirQuick**
This pull request makes `x:expect` reject `@test` at run time if it looks like a mistake. The change is twofold:

- Promote a warning (implemented in #309) to a fatal error.

- Implement what Jeni suggested [10+ years ago](https://github.com/expath/xspec/issues/5#issuecomment-110043244):

  > Perhaps if the result of the test isn't a boolean, there ought to be a requirement for the select attribute and/or content of <x:expect> to be present.

### Summary

| `@test` | <code>@as &#x7C; @href &#x7C; @select &#x7C; child::node()</code> | Current behavior | After this PR |
| --- | --- | --- | --- |
| Exists. An instance of `xs:boolean`. | Exists | <code>@as &#x7C; @href &#x7C; @select &#x7C; child::node()</code> is ignored.<ul><li>XSLT: Warning message (#309)</li><li>XQuery: No warning</li></ul> | Fatal error |
| Exists. Not an instance of `xs:boolean`. | Not exists | `x:expect` is considered to expect an empty sequence | Fatal error |

#### Boolean `@test`

On the current `master`

```xml
<x:expect test="/foo/bar = 3" select="3" />
```

works, although a warning message is emitted.

After this PR, it's a fatal error. To fix it, `@select` should be removed:

```xml
<x:expect test="/foo/bar = 3" />
```

#### Non-boolean `@test`

On the current `master`

```xml
<x:expect test="/foo/bar" />
```

is considered to expect `/foo/bar` to be equal to an empty sequence.

After this PR, it's a fatal error. To fix it, `@test` should be an instance of `xs:boolean`:

```xml
<x:expect test="/foo/bar => empty()" />
```

or `x:expect` should expect an empty sequence explicitly using `@select` or `@as`:

```xml
<x:expect test="/foo/bar" select="()" />
```

```xml
<x:expect test="/foo/bar" as="empty-sequence()" />
```

### Error code

The new fatal error does not have a specific error code. Its error code is a generic one:

- XSLT: [`err:XTMM9000`](https://www.w3.org/TR/xslt-30/#err-XTMM9000)
- XQuery: [`err:FOER0000`](https://www.w3.org/TR/xpath-functions-31/#ERRFOER0000)

See #1074 for the rationale. To reiterate, if we want to reinstate the XSpec-specific error codes, I think it should be done after establishing the facility.

### Commits

- d34649c58b52991174e8c9167472e4019fd5eb78

  The primary change.

- df32f370d28f91eecec371843ef9925f9a378a17

  Deletes or edits some tests to reflect the code change.

- 21c964a1d0fdf7b7f1eb58a7f4de968fa177d341

  Tests `x:expect[empty(@href | @select | @test | child::node())]`. Such an `x:expect` is still valid and considered to expect an empty sequence. (Rejecting such an `x:expect` is not the scope of this pull request.)

- aaf2b7180626e5d02855a53351c2fed2856c8d8d

  Removes some `exactly-one()` and `treat as` from `test/*.xspec`. They are no longer necessary.

### Further work

Update the "Expectations" section of [Wiki](https://github.com/xspec/xspec/wiki/Writing-Scenarios#expectations).

### CODE REVIEW

**Comment by galtm** [comment]
@AirQuick , I started reviewing this and will try to finish by Monday. Do you plan to mention this as a "breaking change" in #1121 ?

**Comment by galtm** [comment]
> Do you plan to mention this as a "breaking change" in #1121 ?

Never mind, I see that you already mentioned this PR there!

**Comment by AirQuick** [comment]
I mentioned there, but somehow I forgot to add a description there.
Now I copied the summary table from this PR to https://github.com/xspec/xspec/issues/1121#issue-668630717.

**Review by galtm — APPROVED** [review]
Looks good, and a major release is a good opportunity to switch from warning to error.

General question about XQuery testing: Is the warning in v1.6 for XSLT tests only, not XQuery tests? If that's true, the "breaking changes" writeup should probably say that. Also, if you think some XQuery test authors might have many instances of this issue, you might even consider adding a warning phase for XQuery (e.g., warning in v2.0 and error in v2.x or v3.0) instead of going straight from no warning to fatal error.

File: src/compiler/base/compile/compile-expect.xsl [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+   <!-- Returns true if x:expect has any comparison factor -->
+   <xsl:function name="x:has-comparison" as="xs:boolean">
+      <xsl:param name="expect" as="element(x:expect)" />
+
+      <xsl:sequence select="$expect/(@href or @select or (node() except x:label))" />
+   </xsl:function>
+
+   <!-- Returns an error string for boolean @test with any comparison factor -->
+   <xsl:function name="x:bad-boolean-test" as="xs:string">
```
[the remark this answers is not in the document]
Comment (AirQuick) [reply src/compiler/base/compile/compile-expect.xsl]: > * x:boolean-with-comparison and x:non-boolean-without-comparison (a little shorter than the above)

I took it in 488262295e2caa9e6a63212d0b48f845fbef4703.

**Comment by AirQuick** [comment]
#1256 has made `@as` available in `x:expect`. So I included `@as` in the comparison factors. (bca62ab74434cb13802195902db705682d3ec3c3)
https://github.com/xspec/xspec/issues/1121#issue-668630717 and the description of this pull request (https://github.com/xspec/xspec/pull/1245#issue-493736793) have been updated to reflect it.

**Comment by AirQuick** [comment]
> General question about XQuery testing: Is the warning in v1.6 for XSLT tests only, not XQuery tests? If that's true, the "breaking changes" writeup should probably say that.

Added `XQuery: No warning` to the summary tables in https://github.com/xspec/xspec/pull/1245#issue-493736793 and https://github.com/xspec/xspec/issues/1121#issue-668630717.

> Also, if you think some XQuery test authors might have many instances of this issue, you might even consider adding a warning phase for XQuery (e.g., warning in v2.0 and error in v2.x or v3.0) instead of going straight from no warning to fatal error.

I think the very initial release (v0.1) of XSpec should have treated `<x:expect test="/foo/bar = 3" select="whatever" />` as a fatal error. So, in terms of boolean `@test`, I would regard this pr as a kind of bug fix and would like to just skip a warning phase for XQuery.
On the other hand, in terms of non-boolean `@test`, this pr is actually a breaking change. Just like  `<x:param />` representing an empty sequence, nothing had been wrong in expecting an empty sequence using `<x:expect test="/foo/bar" />`; but now it's a fatal error.

File: src/compiler/base/compile/compile-expect.xsl [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      <xsl:param name="expect" as="element(x:expect)" />
+
+      <xsl:text expand-text="yes">{name($expect)} has boolean @test, but it also has (@href | @select | child::node()).</xsl:text>
```
[the remark this answers is not in the document]
Comment (AirQuick) [reply src/compiler/base/compile/compile-expect.xsl]: 9ba3fb9e7dfa6334278bdfa88dce7ca5cb54822f includes labels (of not only `x:expect` but also its ancestors &mdash; I found an `x:expect` label alone was often useless) and restores the original phrase.

[the remark this answers is not in the document]
Comment (galtm) [reply src/compiler/base/compile/compile-expect.xsl]: Good point about including ancestry for greater context. I like your solution.

**Review by galtm — APPROVED** [review]
I looked at your latest changes and comments, and they all look good.

**Comment by AirQuick** [comment]
> ### Further work
> Update the "Expectations" section of [Wiki](https://github.com/xspec/xspec/wiki/Writing-Scenarios#expectations).

Just for the record. This has already been done.

**Pull request merged**